python/mead/downloader.py

```diff
@@ -0,0 +1,300 @@
+                json.dump(dcaches, open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json"), "w"), indent=True)
+                return dload_file
+        else:
+            raise RuntimeError("the file {} is not in cache and can not be downloaded".format(file_loc))
+
+
+class DataDownloader(Downloader):
+    def __init__(self, dataset_desc, data_download_cache, enc_dec=False, cache_ignore=False):
+        super(DataDownloader, self).__init__(data_download_cache, cache_ignore)
+        self.dataset_desc = dataset_desc
+        self.data_download_cache = data_download_cache
+        self.enc_dec = enc_dec
+
+    def download(self):
+        import os
+        import json
+        dload_bundle = self.dataset_desc.get("download", None)
+        if dload_bundle is not None:  # download a zip/tar/tar.gz directory, look for train, dev test files inside that.
+            try:
+                dcaches = json.load(open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json")))
+            except IOError:
+                dcaches = {}
+            if dload_bundle in dcaches and check_sanity_dir(dcaches[dload_bundle], self.dataset_desc, self.enc_dec, dload_bundle) and not self.cache_ignore:
+                download_dir = dcaches[dload_bundle]
+                print("files for {} found in cache, not downloading".format(dload_bundle))
+                return {k: os.path.join(download_dir, self.dataset_desc[k]) for k in self.dataset_desc if k.endswith("_file")}
+            else:  # try to download the bundle and unzip
+                url = dload_bundle
+                if not validate_url(url):
+                    raise RuntimeError("can not download from the given url")
+                else:
+                    cache_dir = self.data_download_cache
+                    temp_file = web_downloader(url)
+                    zipd = {'application/gzip': extract_gzip, 'application/zip': extract_zip}
+                    download_dir = extractor(filepath=temp_file, cache_dir=cache_dir, extractor_func=zipd.get(mime_type(temp_file), None))
+                    if "sha1" in self.dataset_desc:
+                        if os.path.split(download_dir)[-1] != self.dataset_desc["sha1"]:
+                            raise RuntimeError("The sha1 of the downloaded file does not match with the provided one")
+                    dcaches.update({url: download_dir})
+                    json.dump(dcaches, open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json"), "w"), indent=True)
+                    return {k: os.path.join(download_dir, self.dataset_desc[k]) for k in self.dataset_desc if k.endswith("_file")}
+        else:  # we have download links to every file or they exist
+            if not self.enc_dec:
+                return {k: SingleFileDownloader(self.dataset_desc[k], self.data_download_cache).download() for k in self.dataset_desc if k.endswith("_file")}
```

Comment:
what is the expected behavior of DataDownloader when SingleFileDownloader fails?

SingleFileDownloader throws an exception, which is not handled.